test/mocking_shell.go

```diff
@@ -17,39 +17,73 @@ import (
 // - Temporarily override certain shell commands with mock implementations.
 //   Temporary mocks are only valid for the next command being run.
 type MockingShell struct {
-	workingDir            string // the directory in which this runner runs
-	homeDir               string // the directory that contains the global Git configuration
-	tempShellOverridesDir string // the directory that stores the mock shell command implementations, ignored if empty
+	workingDir     string // the directory in which this runner runs
+	homeDir        string // the directory that contains the global Git configuration
+	binDir         string // the directory that stores the mock shell command implementations, ignored if empty
+	testOrigin     string // optional content of the GIT_TOWN_REMOTE environment variable
+	hasMockCommand bool   // indicates whether the current test has mocked a command
 }
 
 // NewMockingShell provides a new MockingShell instance that executes in the given directory.
-func NewMockingShell(workingDir string, homeDir string) *MockingShell {
-	return &MockingShell{workingDir: workingDir, homeDir: homeDir}
+func NewMockingShell(workingDir string, homeDir string, binDir string) *MockingShell {
+	return &MockingShell{workingDir: workingDir, homeDir: homeDir, binDir: binDir}
 }
 
-// AddTempShellOverride temporarily mocks the shell command with the given name
-// with the given Bash script.
-func (ms *MockingShell) AddTempShellOverride(name, content string) error {
-	if !ms.hasTempShellOverrides() {
-		err := ms.createTempShellOverridesDir()
-		if err != nil {
-			return fmt.Errorf("cannot create temp shell overrides dir: %w", err)
-		}
+// MockCommand adds a mock for the command with the given name.
+func (ms *MockingShell) MockCommand(name string) error {
+	// create "bin" dir
+	err := os.Mkdir(ms.binDir, 0744)
+	if err != nil {
+		return fmt.Errorf("cannot create mock bin dir: %w", err)
 	}
-	return ioutil.WriteFile(ms.tempShellOverrideFilePath(name), []byte(content), 0744)
+	// write custom "which" command
+	content := fmt.Sprintf("#!/usr/bin/env bash\n\nif [ \"$1\" == %q ]; then\n  echo %q\nelse\n  exit 1\nfi", name, filepath.Join(ms.binDir, name))
```

Comment:
Simplified which command that returns the answer right away without filesystem access.